ui/gtk4.c

```diff
@@ -0,0 +1,2194 @@
+ *   GdkDisplay *dpy;
+ *   gint x_root, y_root;
+ *   if (qemu_input_is_absolute()) {
+ *       return;
+ *   }
+ *   dpy = gtk_widget_get_display(vc->gfx.drawing_area);
+ *   gdk_window_get_root_coords(gtk_widget_get_window(vc->gfx.drawing_area),
+ *                              x, y, &x_root, &y_root);
+ *   gdk_device_warp(gd_get_pointer(dpy),
+ *                   gtk_widget_get_screen(vc->gfx.drawing_area),
+ *                   x_root, y_root);
+ */
+    vc->s->last_x = x;
+    vc->s->last_y = y;
+}
+
+static void gd_cursor_define(DisplayChangeListener *dcl,
+                             QEMUCursor *c)
+{
+    VirtualConsole *vc = container_of(dcl, VirtualConsole, gfx.dcl);
+    GBytes *bytes;
+    GdkTexture *texture;
+    GdkCursor *cursor;
+
+    if (!gtk_widget_get_realized(vc->gfx.drawing_area)) {
+        return;
+    }
+    bytes = g_bytes_new(c->data, c->height * c->width * 4);
+    texture = gdk_memory_texture_new(c->width, c->height,
+                                     GDK_MEMORY_R8G8B8A8,
+                                     bytes, c->width * 4);
+    cursor = gdk_cursor_new_from_texture(texture, c->hot_x, c->hot_y, NULL);
+    gtk_widget_set_cursor(vc->gfx.drawing_area, cursor);
+    g_object_unref(cursor);
+}
+
+static void gd_switch(DisplayChangeListener *dcl,
+                      DisplaySurface *surface)
+{
+    VirtualConsole *vc = container_of(dcl, VirtualConsole, gfx.dcl);
+    bool resized = true;
+
+    trace_gd_switch(vc->label, surface_width(surface), surface_height(surface));
+
+    if (vc->gfx.surface) {
+        cairo_surface_destroy(vc->gfx.surface);
+        vc->gfx.surface = NULL;
+    }
+    if (vc->gfx.convert) {
+        pixman_image_unref(vc->gfx.convert);
+        vc->gfx.convert = NULL;
+    }
+
+    if (vc->gfx.ds &&
+        surface_width(vc->gfx.ds) == surface_width(surface) &&
+        surface_height(vc->gfx.ds) == surface_height(surface)) {
+        resized = false;
+    }
+    vc->gfx.ds = surface;
+
+    if (surface->format == PIXMAN_x8r8g8b8) {
+        /*
+         * PIXMAN_x8r8g8b8 == CAIRO_FORMAT_RGB24
+         *
+         * No need to convert, use surface directly.  Should be the
+         * common case as this is qemu_default_pixelformat(32) too.
+         */
+        vc->gfx.surface = cairo_image_surface_create_for_data
+            (surface_data(surface),
+             CAIRO_FORMAT_RGB24,
+             surface_width(surface),
+             surface_height(surface),
+             surface_stride(surface));
+    } else {
+        /* Must convert surface, use pixman to do it. */
+        vc->gfx.convert = pixman_image_create_bits(PIXMAN_x8r8g8b8,
+                                                   surface_width(surface),
+                                                   surface_height(surface),
+                                                   NULL, 0);
+        vc->gfx.surface = cairo_image_surface_create_for_data
+            ((void *)pixman_image_get_data(vc->gfx.convert),
```

Comment:
This call and the same function called above look weird, having the parenthesis in the next line (at least to me). Is this accepted in the coding guidelines?